lib/puppet/util/character_encoding.rb

```diff
@@ -51,46 +57,39 @@ def convert_to_utf_8(string)
           # UTF-8. We can only guess, so if the string is valid UTF-8 as
           # currently constituted, we default to assuming the string originated
           # in UTF-8 and do not transcode it - we only set external encoding.
-          value_to_encode.force_encoding(Encoding::UTF_8)
-        elsif transcodable?(value_to_encode)
-          # If the string is not currently valid UTF-8 but it can be transcoded,
-          # we can guess this string was not originally unicode. We need it to
-          # be now, so transcode it to UTF-8. For strings with original
-          # encodings like SHIFT_JIS, this should be the final result.
-          value_to_encode.encode!(Encoding::UTF_8)
+          return string.force_encoding(Encoding::UTF_8)
+        elsif currently_valid
+          # If the string is not currently valid UTF-8 but it can be transcoded
+          # (it is valid in its current encoding), we can guess this string was
+          # not originally unicode. Transcode it to UTF-8. For strings with
+          # original encodings like SHIFT_JIS, this should be the final result.
+          return string.encode!(Encoding::UTF_8)
         else
-          # If the string is neither valid UTF-8 as-is nor is transcodable, fail
-          # at this point. It requires user remediation.
+          # If the string is neither valid UTF-8 as-is nor valid in its current
+          # encoding, fail. It requires user remediation.
           raise EncodingError
         end
       rescue EncodingError => detail
-        # catch both our own self-determined failure to transcode as well as any
-        # error on ruby's part, ie Encoding::InvalidByteSequenceError or
-        # Encoding::UndefinedConversionError.
-        raise Puppet::Error, _("#{detail.inspect}: #{value_to_encode.dump} is not valid UTF-8 and cannot be transcoded by Puppet.")
+        # Catch both our own self-determined failure to transcode as well as any
+        # error on ruby's part, ie Encoding::UndefinedConversionError on a
+        # failure to encode!.
+        raise Puppet::Error.new(_("%{error}: %{value} is not valid UTF-8 and cannot be transcoded by Puppet.") %
+          { error: detail.inspect, value: string.dump }, detail)
```

Comment:
sounds good. I really really didn't like doing this raise/catch/re-raise business and it definitely is a huge part of the performance overhead (probably the biggest) but I couldn't think of another way to unify all the various Encoding errors we could raise into a single one with a clear message. I like that option - just don't raise any error at all :) This will certainly be much faster without it.